streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -16,10 +16,298 @@
  */
 package org.apache.kafka.streams.processor.internals;
 
+import org.apache.kafka.clients.consumer.OffsetResetStrategy;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.errors.TopologyException;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder.TopicsInfo;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO KAFKA-12648:
+//  1) synchronize on these methods instead of individual InternalTopologyBuilder methods,
+//  where applicable (ie not used elsewhere in potentially thread-unsafe way)
+//  2) refactor TTD to use TopologyMetadata
+//  3) move implementation of getter-like methods to this class, refactor so InternalTopologyBuilder is just the builder
+//  4) make sure all task-level logging includes NamedTopology
 
 public class TopologyMetadata {
```

Comment:
This is the main work of this PR -- basically each NamedTopology retains its own InternalTopologyBuilder, and the new TopologyMetadata class wraps them all and handles consolidating the metadata, verifying constraints (eg that input topics are unique), and so on. Most of the changes here are just refactoring all the many things that directly depended on the InternalTopologyBuilder to instead go through this class, and then refactoring InternalTopologyBuilder to better fit in the new framework